### ISSUE TITLE
CSRFAction doesn't always tag request with token

### ISSUE THREAD

**Issue by jxtps** [playframework/playframework#7717]
There's a bug in [CSRFActions.scala](https://github.com/playframework/playframework/blob/master/framework/src/play-filters-helpers/src/main/scala/play/filters/csrf/CSRFActions.scala#L63) - if the request may require a CSRF check, but is exempted because: 

````
      if (!csrfActionHelper.requiresCsrfCheck(request)) {
        continue
      } else {
````
then we don't ensure that the request is tagged with a CSRF.Token - it is possible that a request was made to e.g. a +nocsrf endpoint, without a CSRF token in the session. 

Calling CSRF.getToken in such a handler leads to a `No CSRF token was generated for this request! Is the CSRF filter installed?` exception. 

Fix is to have the locally defined `continue` do basically the same things as the end of the if-chain: 

````
      if (csrfActionHelper.getTokenToValidate(request).isEmpty && config.createIfNotFound(request)) {

      // No token in header and we have to create one if not found, so create a new token
      val requestWithNewToken = csrfActionHelper.tagRequestHeaderWithNewToken(request)

      // Once done, add it to the result
      next(requestWithNewToken).map(csrfActionHelper.addTokenToResponse(requestWithNewToken, _))

    } else {
      filterLogger.trace("[CSRF] No check necessary")
      next(request)
    }
````

### Play Version (2.5.x / etc)
2.6.2

### API (Scala / Java / Neither / Both)
Scala

**Comment by gmethvin** [comment]
It actually happens in the else clause there as well. You can see later that after checking the token there is also a `continue`. So it would not work on an action that checked the token then wanted to display a new token. Basically the assumption here is that you would only display token on non-mutative request methods like GET.

I think it should be okay to move the `requiresCsrfCheck` to the enclosing `if` and have that only handle the case that requires a check. So the two cases are mutually exclusive: either we want to check the method for a CSRF token and validate, or we want to tag the request (if not already tagged) and continue.

**Comment by gmethvin** [comment]
@jxtps Also I'm curious what your specific use case is. Why are you rendering a token somewhere you need to use nocsrf. I'd guess that's a POST method.

Like I said I'm not opposed to changing this though. It should simplify the code also. Would gladly accept a PR.

**Comment by jxtps** [comment]
Excellent point, much better solution, PR created. 

Use case: we have a web api that's used in iframes on other sites. Due to the amount of data that needs to be transferred it has to be done by POST => + nocsrf. Due to code sharing with other functionality on the site, a CSRF token was being rendered in the resulting page => exception. 

I doubt that the CSRF checks would reliably work in an iframe context, with third party cookies sometimes getting blocked, but the functionality actually used on that version of the page doesn't need it.

**Comment by jxtps** [comment]
Please hold on merging this - the edit is good, but there's more to the problem.

**Comment by jxtps** [comment]
Ok, so the additional issue was in [CSRFConfig.defaultCreateIfNotFound](https://github.com/playframework/playframework/blob/master/framework/src/play-filters-helpers/src/main/scala/play/filters/csrf/csrf.scala#L85): 
````
  private def defaultCreateIfNotFound(request: RequestHeader) = {
    // If the request isn't accepting HTML, then it won't be rendering a form, so there's no point in generating a
    // CSRF token for it.
    import play.api.http.MimeTypes._
    (request.method == "GET" || request.method == "HEAD") && (request.accepts(HTML) || request.accepts(XHTML))
  }
````
which precluded adding the token to POST requests. 

Unclear if we want to change this default or not as it is relatively straightforward to override by injecting your own config (which can derive from the Play default), but anyone who renders a +nocsrf POST page that needs a CSRF token will be bitten by it.

**Comment by gmethvin** [comment]
@jxtps `defaultCreateIfNotFound` can be overridden by creating a custom binding for the `CSRFConfig`, but it might be nice to be able to do that in the configuration file. We could have something like `play.filters.csrf.addTokenMethods`, and add some additional logic in `CSRFConfig.fromConfiguration`. As long as the default behavior is the same it should be perfectly fine for 2.6.x.

We may want to consider having `createIfNotFound` by default return true for 2.7+, with an option to filter specific methods and content types. The main reason being for simplicity. For strict bodies we already do a check to see if the token has been used before printing it out, so I think for the vast majority of cases `createIfNotFound` doesn't provide much of an optimization.

It's also worth mentioning that if you want to force the CSRF token to be added on specific actions, you can use the `CSRFAddToken` action builder.

**Comment by jxtps** [comment]
So this is going to be a little of a tangent, but I ended up rolling my own CSRF filtering. 

Yes, that feels like a bad idea, but there was so many configuration options / things to maybe customize, with so many layers / participants (CSRFFilter, CSRFAction CSRFConfig, defaultThisThatAndTheOther, token signers, providers, error handlers, helpers, etc), and I had a hard time getting it to work well: requests were getting blocked that probably shouldn't have gotten blocked.  

The basic observation is that a CSRF token is needed if and only if there's a session that privileges the action, as otherwise the attacker can just issue the request themselves, there's no need to bother with CSRF. (Yes, the authentication header is another possibility, but we only use that for API requests). 

For unclear reasons, we're getting uses who get a CSRF token set on the page (= in the form), but then submit requests without a token in their session (= in the request header). I believe this is because the user's browser refused to set the session cookie (plugin / security setting / whatever), but that should not lead to a CSRF blocking of the request (though it will cause other problems, but they should be dealt with separately). 

We're also getting requests with mis-matched page-vs-session tokens. Both tokens are present, they're just different. I don't think anyone is currently hacking our site, so this is really weird. It's like they rendered a page with one token, then got a new CSRF session token (= signed out, then got a new one set), then submitted a form on the first page. Weird usage, but it's happening and I don't think it's malicious use, just overzealous CSRF blocking. 

So what I ended up doing was to roll my own CSRF token by signing the userId that I store in the session + a timestamp, and then pickling that to a string with the timestamp and the signature. Changes on every request (= no BREACH), but they can be independently verified as long as the userId doesn't change, and a user has a header token if and only if there's a session, and if there isn't then there's nothing to block. 

The only scenario where this blocks a somewhat legitimate request is if a user signs in under one account, renders a page, then opens up the site in a new browser window, signs out and signs in under another account, then submits the form on the first page. I'll live with that. 

I'm not sure if there's any takeaway from this - the key difference that I couldn't configure my way out of was the definition of the token, that it's session === token, and that the current code uses `tokenProvider.compareTokens(headerToken, queryStringToken)` - in what I'm describing that would have to be `tokenProvier.compareTokens(requestHeader, queryStringToken)`. 

Regarding the logging of the warning messages, we probably want to change that to a utility method on the CSRFAction class that includes the URI, the HTTP referrer, and the header token - turns out all three were quite useful in the end. 

Separately, it would be very helpful for people needing minor local Play modifications to not have so much of Play's methods and variables be private. In order to change the `CSRFAction`'s behavior pending any updates to the framework, I had to create a full copy-paste of the class in `play.filters.csrf` as otherwise it wasn't possible to access the functionality needed. If the default were to use protected instead of private it would make things much more amenable to re-use. Encapsulation and state hiding are great, but so are extensibility & reuse.

**Comment by gmethvin** [comment]
> For unclear reasons, we're getting uses who get a CSRF token set on the page (= in the form), but then submit requests without a token in their session (= in the request header). I believe this is because the user's browser refused to set the session cookie (plugin / security setting / whatever), but that should not lead to a CSRF blocking of the request (though it will cause other problems, but they should be dealt with separately).

If the browser refused to set the session cookie, then the updated session cookie would not be passed back to the server, so there is no token in the session for the filter to compare it to. Maybe I misunderstand what you're saying though.

> Separately, it would be very helpful for people needing minor local Play modifications to not have so much of Play's methods and variables be private.

There is a tradeoff here, though. If we make methods public/protected and part of the official API, we have to worry about breaking everyone's custom implementations when we update APIs. Most of these methods are really not designed to be part of the public API, and I suspect if they were they would have been designed differently. The current implementation is sort of a mess. If people copy/paste the code and make their own modifications, there is no need to worry about that. If there are specific parts you think should be more configurable we're happy to consider though.

Considering you've solved the problem yourself, let us know if you have an interest in continuing your PR in #7720.

**Comment by jxtps** [comment]
> so there is no token in the session for the filter to compare it to.

Yes, that is precisely what happens, and the request gets blocked with `Check failed because no token found in headers.`

Public vs private: fair point. 

Continuing the PR: the edit as it stands should probably be merged for better code clarity, but as noted it would not independently pass any new tests added to cover the missing token case. I'm afraid my home-brew solution isn't general enough to merit sharing (which I would otherwise happily do).

**Comment by philipwhiuk** [comment]
Adding @jxtps - it would be nice if there was a good way to add extra detail to stuff like this - e.g. the origin IP of the requester not providing the CSRF token

https://github.com/playframework/playframework/blob/4b97fa0af1652d8a5b10c12d0e2b419ea1e6ccf1/web/play-filters-helpers/src/main/scala/play/filters/csrf/CSRFActions.scala#L88

I'm coming up against the same problem and debating copying @jxtps and rolling my own CSRF implementation to debug it.
